### PR TITLE
[Support RP] CLI release GA version 2024-04-01

### PR DESCRIPTION
**Pull request by RudraSharma93Microsoft**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally? (`pip install wheel==0.30.0` required)
- [x] My extension version conforms to the [Extension version schema](https://github.com/Azure/azure-cli/blob/release/doc/extensions/versioning_guidelines.md)

For new extensions:

- [x] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
Thank you for your contribution! We will review the pull request and get back to you soon.

File: src/support/azext_support/commands.py [hunk 1]
```diff
@@ -3,53 +3,54 @@
...
+        loader=self
+    )
+    self.command_table["support no-subscription tickets list"] = TicketListNoSubscription(
+        loader=self
+    )
+    with self.command_group("support no-subscription file") as g:
+        g.custom_command("upload", "upload_files_no_subscription")
+
+    with self.command_group("support in-subscription file") as g:
+        g.custom_command("upload", "upload_files_in_subscription")
```
Comment (kairu-ms) [inline src/support/azext_support/commands.py:49]: Please add help in _help.py file for those two commands.

**Review by Moazzem-Hossain-pixel — APPROVED** [review]
Approved with thanks

**Comment by yonzhan** [comment]
Please fix CI issues

**Review by Moazzem-Hossain-pixel — APPROVED** [review]
Thanks for contributions.

**Review by Moazzem-Hossain-pixel — CHANGES_REQUESTED** [review]
Thanks

**Review by Moazzem-Hossain-pixel — APPROVED** [review]
Gladly approved.

File: src/support/azext_support/commands.py (hunk 1 above)
Comment (AllyW) [reply src/support/azext_support/commands.py:49]: Also provide help message for related parameters in _params.py

File: src/support/azext_support/__init__.py [hunk 2]
```diff
@@ -1,32 +1,33 @@
-# --------------------------------------------------------------------------------------------
-# Copyright (c) Microsoft Corporation. All rights reserved.
-# Licensed under the MIT License. See License.txt in the project root for license information.
-# --------------------------------------------------------------------------------------------
-
-from azure.cli.core import AzCommandsLoader
-
-from azext_support._help import helps  # pylint: disable=unused-import
```
Comment (kairu-ms) [inline src/support/azext_support/__init__.py:8]: Please add back this line

File: src/support/azext_support/__init__.py [hunk 3]
```diff
@@ -1,32 +1,33 @@
...
-                         custom_command_type=support_custom)
-
-    def load_command_table(self, args):
-        from azext_support.commands import load_command_table
-        load_command_table(self, args)
-        return self.command_table
-
-    def load_arguments(self, command):
-        from azext_support._params import load_arguments
-        load_arguments(self, command)
```
Comment (kairu-ms) [inline src/support/azext_support/__init__.py:29]: please add back this as well

File: src/support/azext_support/_params.py [hunk 4]
```diff
@@ -1,155 +0,0 @@
...
-                                       get_supported_languages_for_update,
-                                       get_supported_timezones_for_create,
-                                       get_supported_timezones_for_update)
-from azext_support._validators import datetime_type
-from azure.cli.core.commands.parameters import (get_enum_type,
-                                                get_three_state_flag)
-from knack.arguments import CLIArgumentType
-
-
-def load_arguments(self, _):
```
[the remark this answers is not in the document]
Comment (RudraSharma93Microsoft) [reply src/support/azext_support/_params.py:16]: Done

File: src/support/azext_support/__init__.py (hunk 3 above)
Comment (RudraSharma93Microsoft) [reply src/support/azext_support/__init__.py:29]: Done

File: src/support/azext_support/__init__.py (hunk 2 above)
Comment (RudraSharma93Microsoft) [reply src/support/azext_support/__init__.py:8]: Done

File: src/support/azext_support/commands.py (hunk 1 above)
Comment (RudraSharma93Microsoft) [reply src/support/azext_support/commands.py:49]: Done

File: src/support/azext_support/azext_metadata.json [hunk 5]
```diff
@@ -1,4 +1,4 @@
-{
-    "azext.isPreview": false,
-    "azext.minCliCoreVersion": "2.0.81"
+{
+    "azext.isPreview": false,
```
Comment (AllyW) [inline src/support/azext_support/azext_metadata.json:2]: if not preview, this line can be deleted.

Comment (AllyW) [inline src/support/azext_support/azext_metadata.json:2]:
```suggestion

```

File: src/support/azext_support/_utils.py [hunk 6]
```diff
@@ -43,15 +45,70 @@ def parse_support_area_path(problem_classification_id):
...
-            unauthorized_error
+    upload_input = {
+        "file_name": full_file_name,
+        "file_workspace_name": file_workspace_name,
+        "chunk_index": chunk_index,
+        "content": string_encoded_content,
+    }
 
-    return "Bearer " + creds[1]
+    Upload(cli_ctx=cmd.cli_ctx)(command_args=upload_input)
```
Comment (AllyW) [inline src/support/azext_support/_utils.py:113]: There are number_of_chunks file piece and only upload the last piece?

File: src/support/azext_support/_validators.py [hunk 7]
```diff
@@ -87,3 +113,79 @@ def _is_guid(guid):
...
+        raise CLIError(resp["message"])
+
+
+def _validate_file_path(file_path):
+    if not os.path.exists(file_path):
+        raise CLIError("File does not exist!")
+
+
+def _validate_file_name(file_name):
+    if len(file_name) > max_file_name_length:
```
Comment (AllyW) [inline src/support/azext_support/_validators.py:180]: Const var recommend set to be all uppercase

Comment (AllyW) [reply src/support/azext_support/_validators.py:180]: same as following

File: src/support/azext_support/custom.py [hunk 8]
```diff
@@ -6,174 +6,276 @@
...
+
+    def _cli_arguments_loader(self):
+        args = super()._cli_arguments_loader()
+        return set_configured_defaults(args)
+
+    def pre_operations(self):
+        from azext_support._validators import _validate_tickets_create
+        from azure.cli.core.aaz import AAZUndefined
+
+        super().pre_operations()
```
Comment (AllyW) [inline src/support/azext_support/custom.py:77]: rewrote pre_operation don't need and should not call its original method.
followings are the same

File: src/support/azext_support/custom.py [hunk 9]
```diff
@@ -6,174 +6,276 @@
...
+        args = self.ctx.args
+        if args.technical_resource != AAZUndefined:
```
Comment (AllyW) [inline src/support/azext_support/custom.py:79]: function has_value is what we use to justify whether aaz cmd arg is empty

File: src/support/azext_support/custom.py [hunk 10]
```diff
@@ -6,174 +6,276 @@
...
+            _validate_tickets_create(
+                self.cli_ctx,
+                str(args.problem_classification),
```
Comment (AllyW) [inline src/support/azext_support/custom.py:82]: And `to_serialized_data()` is how we fetch the raw value from aaz cmd args.
Followings are the same

File: src/support/azext_support/custom.py [hunk 11]
```diff
@@ -6,174 +6,276 @@
...
+
+    class SupportTicketsList(_List.SupportTicketsList):
+
+        @property
+        def query_parameters(self):
+            from azure.cli.core.aaz import AAZUndefined
+
+            parameters = super().query_parameters
+            args = self.ctx.args
+            if args.filter == AAZUndefined and args.pagination_limit == AAZUndefined:
```
Comment (AllyW) [inline src/support/azext_support/custom.py:169]: Adjust query arg should be done in rewrote `pre_operation` through customerized  value assignment. 
you can refer to code here: https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/network/operations/dns.py#L829-L832

Comment (AllyW) [reply src/support/azext_support/custom.py:169]: Followings are the same

**Comment by AllyW** [comment]
> ❌Azure CLI Extensions Breaking Change Test
> > ❌support
> > 
> > 
> > 
> > rule
> > cmd_name
> > rule_message
> > suggest_message
> > 
> > 
> > 
> > 
> > ❌ [1012 - SubgroupRemove](https://github.com/Azure/azure-cli/blob/dev/doc/breaking_change_rules/1012.md)
> > support tickets
> > sub group `support tickets` removed
> > please confirm sub group `support tickets` removed
> > 
> > 
> > ⚠️ [1011 - SubgroupAdd](https://github.com/Azure/azure-cli/blob/dev/doc/breaking_change_rules/1011.md)
> > support in-subscription
> > sub group `support in-subscription` added
> > 
> > 
> > 
> > ⚠️ [1011 - SubgroupAdd](https://github.com/Azure/azure-cli/blob/dev/doc/breaking_change_rules/1011.md)
> > support no-subscription
> > sub group `support no-subscription` added
> > 
> > 
> > 
> > ⚠️ [1008 - ParaPropAdd](https://github.com/Azure/azure-cli/blob/dev/doc/breaking_change_rules/1008.md)
> > support services problem-classifications list
> > cmd `support services problem-classifications list` update parameter `service_name`: added property `aaz_type=string`
> > 
> > 
> > 
> > ⚠️ [1008 - ParaPropAdd](https://github.com/Azure/azure-cli/blob/dev/doc/breaking_change_rules/1008.md)
> > support services problem-classifications list
> > cmd `support services problem-classifications list` update parameter `service_name`: added property `type=string`
> > 
> > 
> > 
> > ⚠️ [1008 - ParaPropAdd](https://github.com/Azure/azure-cli/blob/dev/doc/breaking_change_rules/1008.md)
> > support services problem-classifications show
> > cmd `support services problem-classifications show` update parameter `problem_classification_name`: added property `aaz_type=string`
> > 
> > 
> > 
> > ⚠️ [1008 - ParaPropAdd](https://github.com/Azure/azure-cli/blob/dev/doc/breaking_change_rules/1008.md)
> > support services problem-classifications show
> > cmd `support services problem-classifications show` update parameter `problem_classification_name`: added property `type=string`
> > 
> > 
> > 
> > ⚠️ [1008 - ParaPropAdd](https://github.com/Azure/azure-cli/blob/dev/doc/breaking_change_rules/1008.md)
> > support services problem-classifications show
> > cmd `support services problem-classifications show` update parameter `service_name`: added property `aaz_type=string`
> > 
> > 
> > 
> > ⚠️ [1008 - ParaPropAdd](https://github.com/Azure/azure-cli/blob/dev/doc/breaking_change_rules/1008.md)
> > support services problem-classifications show
> > cmd `support services problem-classifications show` update parameter `service_name`: added property `type=string`
> > 
> > 
> > 
> > ⚠️ [1008 - ParaPropAdd](https://github.com/Azure/azure-cli/blob/dev/doc/breaking_change_rules/1008.md)
> > support services show
> > cmd `support services show` update parameter `service_name`: added property `aaz_type=string`
> > 
> > 
> > 
> > ⚠️ [1008 - ParaPropAdd](https://github.com/Azure/azure-cli/blob/dev/doc/breaking_change_rules/1008.md)
> > support services show
> > cmd `support services show` update parameter `service_name`: added property `type=string`

Moreover, please at lease log the subgroup changes into HISTORY.rst file. Current log message is too general

File: src/support/HISTORY.md [hunk 12]
```diff
@@ -1,6 +1,10 @@
 Release History
 ===============
 
+2.0.0
+-----
+* Upgrade CLI Support extension to new Support RP API 2024-04-01
```
Comment (AllyW) [inline src/support/HISTORY.md:6]: Current update log is rather general, please log the subgroup level change into HISTORY notes.
image

**Review by Moazzem-Hossain-pixel — APPROVED** [review]
Thanks for favor.

**Review by AllyW — APPROVED** [review]

**Pull request merged**